data_iterator.go

```diff
@@ -107,81 +130,135 @@ func (d *DataIterator) Run(tables []*TableSchema) {
 						d.ErrorHandler.Fatal("inline_verifier", err)
 					default:
-						logger.WithError(err).Error("failed to iterate table")
+						d.logger.WithError(err).Error("failed to iterate table")
 						d.ErrorHandler.Fatal("data_iterator", err)
 					}
 
 				}
-
-				logger.Debug("table iteration completed")
-
-				// Right now the BatchWriter.WriteRowBatch happens synchronously in
-				// this method. If it ever becomes async, this MarkTableAsCompleted
-				// call MUST be done in WriteRowBatch somehow.
-				d.StateTracker.MarkTableAsCompleted(table.String())
 			}
 		}()
 	}
 
-	i := 0
-	loggingIncrement := len(tablesWithData) / 50
-	if loggingIncrement == 0 {
-		loggingIncrement = 1
-	}
+	for _, key := range d.paginationKeys {
+		tableName := key.table.String()
+		stateBatches, loadBatchesFromState := d.StateTracker.batchProgress[tableName]
 
-	for table, _ := range tablesWithData {
-		tablesQueue <- table
-		i++
-		if i%loggingIncrement == 0 {
-			d.logger.WithField("table", table.String()).Infof("queued table for processing (%d/%d)", i, len(tablesWithData))
+		if loadBatchesFromState {
+			for _, batch := range stateBatches {
+				if batch.Completed {
+					continue
+				}
+
+				batchQueue <- &PaginationKey{
+					table:              key.table,
+					startPaginationKey: batch.LatestPaginationKey,
+					endPaginationKey:   batch.EndPaginationKey,
+				}
+			}
+		} else {
+			// Set start to minus one since cursor is searching for greater values
+			tableStartPaginationKey := key.startPaginationKey - 1
+			tableEndPaginationKey := key.endPaginationKey
+
+			// Number of batches are set to number of processes, unless each batch becomes smaller than the cursor size
+			keyInterval := tableEndPaginationKey - tableStartPaginationKey
+			concurrencyBatchSize := math.Ceil(float64(keyInterval) / float64(d.Concurrency))
+			batchSize := uint64(math.Max(concurrencyBatchSize, float64(d.CursorConfig.BatchSize)))
+
+			d.logger.WithFields(logrus.Fields{
+				"table":              tableName,
+				"batchSize":          batchSize,
+				"endPaginationKey":   tableEndPaginationKey,
+				"startPaginationKey": tableStartPaginationKey,
+			}).Debugf("queueing %d batches", (keyInterval/batchSize)+1)
+
+			for batchStartPaginationKey := tableStartPaginationKey; batchStartPaginationKey < tableEndPaginationKey; batchStartPaginationKey += batchSize {
+				batchEndPaginationKey := batchStartPaginationKey + batchSize
+
+				// Set batchEndPaginationKey to endPaginationKey if out of bounds
+				if batchEndPaginationKey > tableEndPaginationKey {
+					batchEndPaginationKey = tableEndPaginationKey
+				}
+
+				paginationKey := &PaginationKey{
+					table:              key.table,
+					startPaginationKey: batchStartPaginationKey,
+					endPaginationKey:   batchEndPaginationKey,
+				}
+
+				d.StateTracker.RegisterBatch(tableName, paginationKey.BatchId(), paginationKey.startPaginationKey, paginationKey.endPaginationKey)
```

Comment:
would be great to see a log after the batch iteration regarding how many batches are created per-table

Reply:
Do you mean more verbose than https://github.com/Shopify/ghostferry/blame/batch-data-iterator/data_iterator.go#L191?